### ISSUE TITLE
Trying to implement DropConnect (Wan et. al, 2013) in PyTorch 1.5

### ISSUE THREAD

**Issue by jna29** [pytorch/pytorch#40193]
## 🐛 Bug

Applying the `WeightDrop` layer to an `LSTM` object results in a warning saying that the RNN weights are not in a contiguous chunk (see below). I believe this to be related to the fact that we are replacing a `nn.Parameter` of the LSTM with a `torch.Tensor` which is then written to a different area in memory. 

I am not sure how to resolve this.

This happens on PyTorch 1.5 but also on 1.4 and 1.3.

I believe #32563 and #32346 are related.

## To Reproduce

A minimal code example is given below. Oddly enough, this only seems to happen in training mode (`self.training=True`).

`weight_drop.py` (Copied and modified from https://github.com/salesforce/awd-lstm-lm) 

```
import torch
from torch.nn import Parameter

class WeightDrop(torch.nn.Module):
    def __init__(self, module, weights, dropout=0, variational=False):
        super(WeightDrop, self).__init__()
        self.module = module
        self.weights = weights
        self.dropout = dropout
        self.variational = variational
        self._setup()

    def widget_demagnetizer_y2k_edition(*args, **kwargs):
        # We need to replace flatten_parameters with a nothing function
        # It must be a function rather than a lambda as otherwise pickling explodes
        # We can't write boring code though, so ... WIDGET DEMAGNETIZER Y2K EDITION!
        # (╯°□°）╯︵ ┻━┻
        return

    def _setup(self):
        # Terrible temporary solution to an issue regarding compacting weights re: CUDNN RNN
        if issubclass(type(self.module), torch.nn.RNNBase):
            self.module.flatten_parameters = self.widget_demagnetizer_y2k_edition

        for name_w in self.weights:
            print('Applying weight drop of {} to {}'.format(self.dropout, name_w))
            w = getattr(self.module, name_w)
            del self.module._parameters[name_w]
            self.module.register_parameter(name_w + '_raw', Parameter(w.data))

    def _setweights(self):
        for name_w in self.weights:
            raw_w = getattr(self.module, name_w + '_raw')
            w = None
            if self.variational:
                mask = torch.autograd.Variable(torch.ones(raw_w.size(0), 1))
                if raw_w.is_cuda: mask = mask.cuda()
                mask = torch.nn.functional.dropout(mask, p=self.dropout, training=True)
                w = mask.expand_as(raw_w) * raw_w
            else:
                w = torch.nn.functional.dropout(raw_w, p=self.dropout, training=self.training)

            # This is because we may call this function in non-training mode first and so, as self.training=False, w is
            # a nn.Parameter and thus self.module.weight remains a Parameter of self.module when we don't want it to.
            if name_w in self.module._parameters:
                del self.module._parameters[name_w]
            setattr(self.module, name_w, w)

    def forward(self, *args):
        self._setweights()
        return self.module.forward(*args)
```
```
import torch.nn as nn

input_size, batch_size, hidden_size, seq_len = 5, 20, 100, 50
drop_prob = 0.5

device = torch.device('cuda')

model = nn.LSTM(input_size, hidden_size, num_layers=1, bias=True, dropout=0).to(device)
model_wd = WeightDrop(model, ['weight_hh_l0'], dropout=drop_prob)

inputs = torch.randn(seq_len, batch_size, input_size).to(device)
hidden_and_cell = (torch.zeros(1, batch_size, hidden_size, device=device), torch.zeros(1, batch_size, hidden_size, device=device))

model_wd.train()

_ = model_wd(inputs, hidden_and_cell)
### /pytorch/aten/src/ATen/native/cudnn/RNN.cpp:1269: UserWarning: RNN module weights are not part of single contiguous chunk of memory. This means they need to be compacted at every call, possibly greatly increasing memory usage. To compact weights again call flatten_parameters().

```
Note that the values chosen for the parameters are arbitrary and the problem should persist with any setting of these values.

## Environment

Result of `https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py`

Collecting environment information...
PyTorch version: 1.5.0+cu101
Is debug build: No
CUDA used to build PyTorch: 10.1

OS: Springdale Linux 7.7 (Verona)
GCC version: (GCC) 4.8.5 20150623 (Red Hat 4.8.5-39)
CMake version: version 2.8.12.2

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: Tesla K80
Nvidia driver version: 418.67
cuDNN version: Could not collect

Versions of relevant libraries:
[pip3] numpy==1.18.4
[pip3] torch==1.5.0+cu101
[pip3] torchvision==0.6.0+cu101
[conda] Could not collect

## Additional context

cc @albanD @mruberry

**Comment by vincentqb** [comment]
Have you tried adding `self.module.flatten_parameters()` as follows? Can you also copy the exact error that you see?
```
def forward(self, *args):
        self._setweights()
        self.module.flatten_parameters()
        return self.module.forward(*args)
```

**Comment by jna29** [comment]
@vincentqb The problem still persists because `self.module.flatten_parameters` has been overridden with `widget_demagnetizer_y2k_edition` (defined above) which does nothing.

However, after reversing the override and putting `self.module.flatten_parameters` after `self._setweights` it seems to work now. 

I thought I had tried this before but what I had done was put `self.module.flatten_parameters` before `self._setweights`.

Thanks for the help.

**Comment by vincentqb** [comment]
Glad to know this worked :) I'll close the issue, but please feel free to re-open if it persits.
